Xamarin.PropertyEditing.Windows/ChoiceControl.cs

```diff
@@ -1,58 +1,108 @@
 ﻿using System;
-using System.Collections;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
 using System.Windows;
 using System.Windows.Controls;
+using System.Windows.Controls.Primitives;
+using System.Windows.Media;
 
 namespace Xamarin.PropertyEditing.Windows
 {
-	[TemplatePart (Name = "list", Type = typeof(ListBox))]
-	internal class ChoiceControl
-		: Control
+	internal class ChoiceItem
+		: NotifyingObject
 	{
-		public event EventHandler SelectedItemChanged;
+		public string Name
+		{
+			get { return this.name; }
+			set
+			{
+				if (this.name == value)
+					return;
 
-		public static readonly DependencyProperty ItemsSourceProperty = DependencyProperty.Register ("ItemsSource", typeof (IEnumerable), typeof (ChoiceControl));
+				this.name = value;
+				OnPropertyChanged();
+			}
+		}
 
-		public IEnumerable ItemsSource
+		public string Tooltip
 		{
-			get { return (IEnumerable)GetValue (ItemsSourceProperty); }
-			set { SetValue (ItemsSourceProperty, value); }
-		}
+			get { return this.tooltip; }
+			set
+			{
+				if (this.tooltip == value)
+					return;
 
-		public static readonly DependencyProperty ItemTemplateProperty = DependencyProperty.Register ("ItemTemplate", typeof (DataTemplate), typeof (ChoiceControl));
+				this.tooltip = value;
+				OnPropertyChanged();
+			}
+		}
 
-		public DataTemplate ItemTemplate
+		public object Value
 		{
-			get { return (DataTemplate)GetValue (ItemTemplateProperty); }
-			set { SetValue (ItemTemplateProperty, value); }
+			get { return this.value; }
+			set
+			{
+				if (this.value == value)
+					return;
+
+				this.value = value;
+				OnPropertyChanged();
+			}
 		}
 
-		public static readonly DependencyProperty SelectedItemProperty = DependencyProperty.Register ("SelectedItem", typeof (object), typeof (ChoiceControl));
+		private object value;
+		private string tooltip;
+		private string name;
+	}
 
-		public object SelectedItem
-		{
-			get { return GetValue (SelectedItemProperty); }
-			set { SetValue (SelectedItemProperty, value); }
-		}
+	[TemplatePart (Name = "list", Type = typeof(ListBox))]
+	internal class ChoiceControl
+		: Selector
+	{
+		public event EventHandler SelectedItemChanged;
 
 		public override void OnApplyTemplate ()
 		{
 			base.OnApplyTemplate ();
 
-			this.list = (ListBox)GetTemplateChild ("list");
-			this.list.SelectionChanged += OnListSelectionChanged;
+			ItemContainerGenerator.StatusChanged += (sender, args) => {
```

Comment:
I assume we don't need to worry about the template changing, thus causing this method to be invoked twice and two event handlers being attached?